### PR TITLE
Add support for asynchronous evaluation of QNode collections

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

Currently, QNode collections use a simple for loop to evaluate each contained QNode.

**Description of the Change:**

* Uses Dask to parallelize the evaluation of qnode collections.

  For example, using `qml.map` to create a collection of QNodes broadcast over two devices and two different observables,

  ```python
  >>> qpu1 = qml.device("forest.qvm", device="Aspen-4-4Q-D", shots=1000)
  >>> qpu2 = qml.device("forest.qvm", device="Aspen-7-4Q-B", shots=1000)
  >>> obs_list = [qml.PauliX(0), qml.PauliZ(0) @ qml.PauliZ(1)]
  >>> qnodes = qml.map(qml.templates.StronglyEntanglingLayers, obs_list, [qpu1, qpu2])
  ```
  We can now time the serial and parallel modes:
  ```python
  >>> %timeit qnodes(params)
  5.16 s ± 162 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
  >>> %timeit qnodes(params, parallel=True)
  2.99 s ± 40.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
  ```

* `test_collections.py` was getting large and unweildy, so moved them into a `tests/collections` folder

**Benefits:**

* For simulators and devices living outside of Python, this leads to a significant speedup in evaluation.

**Possible Drawbacks:**

* Gradient calculation also works asynchronously for both Autograd and PyTorch. It **doesn't** work for TensorFlow; with `parallel=True`, the gradient tape just returns `None`.

* This **should not be used** with `default.qubit`, as it was not written for parallel computation and there may be race conditions. It should instead be used for external hardware devices (IBMQ, QCS) and external simulators (like the QVM, qiskit aer).

**Related GitHub Issues:** n/a

### CODE REVIEW

File: pennylane/collections.py [hunk 1]
```diff
@@ -467,9 +520,29 @@ def evaluate(self, args, kwargs):
             list: the results from each QNode
         """
         results = []
+        _async = kwargs.pop("parallel", False)
```
Comment (johannesjmeyer) [inline pennylane/collections.py]: I would consider renaming this to `_parallel` so that the name is in direct correspondence with the kwarg name.

Comment (josh146) [reply pennylane/collections.py]: Done!

**Comment by josh146** [comment]
> But default.qubit should not be a problem if every thread has its own copy of default.qubit, should it?

That is what I thought as well. I tried manually making deep copies of default.qubit to see what would happen, but it seems like a race condition still occurs somewhere. Note that I can only recreate this in a special case: when there are two default qubit devices and > 2 observables, and the two devices are 'interleaved'. Repeated calls to `qnodes(params, parallel=True)` will result in stochastic results.

File: pennylane/collections.py [hunk 2]
```diff
@@ -421,6 +422,58 @@ def qnode2(x, y):
...
+    .. warning::
+
+        Asynchronous evaluation is experimental --- please report all bugs and issues
+        to our GitHub page. It currently works with all interfaces, however backpropagation
+        and gradient computation is limited to Autograd and PyTorch. **Quantum gradients
+        using TensorFlow in asynchronous mode is currently not supported**.
+
+        You will find the best speedups when using asynchronous mode when QNodes are to
+        be evaluated on external hardware devices or external simulators. **It is not
+        advised to use asynchronous mode with** ``default.qubit``**.**
```
Comment (trbromley) [inline pennylane/collections.py]: We could emphasise even more and say "Do not use asynchronous mode with `default.qubit`"

File: pennylane/collections.py [hunk 3]
```diff
@@ -467,9 +520,29 @@ def evaluate(self, args, kwargs):
...
+        parallel = kwargs.pop("parallel", False)
+        _scheduler = kwargs.pop("scheduler", "threads")
+
+        if parallel:
+            try:
+                import dask
+            except:
+                raise ImportError("Dask must be installed for parallel evaluation. "
+                                  "\nDask can be installed using pip:"
+                                  "\n\npip install dask[delayed]")
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/collections.py]: I didn't have any problem with
`pip install dask[delayed]`

File: pennylane/collections.py [hunk 4]
```diff
@@ -467,9 +520,29 @@ def evaluate(self, args, kwargs):
             list: the results from each QNode
         """
         results = []
+        parallel = kwargs.pop("parallel", False)
+        _scheduler = kwargs.pop("scheduler", "threads")
```
Comment (trbromley) [inline pennylane/collections.py]: Is there a case where we wouldn't want the scheduler to default to `threads`? I think it's fine but just wondering 🤔

File: requirements.txt [hunk 5]
```diff
@@ -6,3 +6,4 @@ autograd
 toml
 appdirs
 semantic_version==2.6
+dask[delayed]
```
Comment (trbromley) [inline requirements.txt]:
```suggestion
dask[delayed]==2.9.2
```
Worth continuing the version pinning convention?

File: pennylane/collections.py [hunk 6]
```diff
@@ -102,6 +102,7 @@ def my_template(params, wires, **kwargs):
     if not isinstance(device, Sequence):
         # broadcast the single device over all observables
         device = [device] * len(observables)
+        # device = [copy.deepcopy(device) for _ in range(len(observables))]
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/collections.py]:
```suggestion
```

File: pennylane/collections.py [hunk 7]
```diff
@@ -421,6 +422,58 @@ def qnode2(x, y):
 
     where the results from each QNode have been flattened and concatenated
     into a single one-dimensional list.
+
+    .. caution::
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/collections.py]: The main reason for having two is so that both would stand out (both are equally important for the users to see). Having one large box would maybe 'drown out' the warnings more.

File: pennylane/collections.py (hunk 4 above)
Comment (josh146) [reply pennylane/collections.py]: Don't think so! `processes` is slower and has more overhead, and `Client` only makes sense in a distributed set up

File: requirements.txt (hunk 5 above)
Comment (josh146) [reply requirements.txt]: Good idea!

File: requirements.txt [hunk 8]
```diff
@@ -6,3 +6,4 @@ autograd
 toml
 appdirs
 semantic_version==2.6
+dask[delayed]==2.9.2
```
Comment (josh146) [inline requirements.txt]: Looks like we need to revert this for the TensorFlow 1.X tests to pass :(
```suggestion
dask[delayed]
```

File: pennylane/collections.py [hunk 9]
```diff
@@ -421,6 +422,58 @@ def qnode2(x, y):
...
+        >>> %timeit qnodes(params, parallel=True)
+        2.99 s ± 40.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+
+
+    .. warning::
+
+        Asynchronous evaluation is experimental --- please report all bugs and issues
+        to our GitHub page. It currently works with all interfaces, however backpropagation
+        and gradient computation is limited to Autograd and PyTorch. **Quantum gradients
+        using TensorFlow in asynchronous mode is currently not supported**.
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/collections.py]: I think this should be 'is', no? :thinking: I'm actually not sure, 'are' used to sound wrong in my head but now it's starting to change. @trbromley?

File: pennylane/collections.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/collections.py]: :laughing:

File: pennylane/collections.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/collections.py]: That's odd, `dask[delayed]` works for me as well (and is the installation instruction they have on their website) :thinking:

File: tests/collections/test_qnode_collection.py [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+        dev = qml.device("default.qubit", wires=1)
+        qnodes = [
+            qml.QNode(circuit, dev, interface="autograd"),
+            qml.QNode(circuit, dev, interface=None),
+        ]
+
+        with pytest.raises(ValueError, match="do not all use the same interface"):
+            qc.extend(qnodes)
+
+    def test_extend_interface_mistmatch(self):
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/collections/test_qnode_collection.py]: Oops!

File: tests/collections/test_qnode_collection.py [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+        interface to the QNode collection are appended"""
+        qc = qml.QNodeCollection()
+
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        dev = qml.device("default.qubit", wires=1)
+        qnode1 = qml.QNode(circuit, dev, interface="autograd")
+        qnode2 = qml.QNode(circuit, dev, interface=None)
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/collections/test_qnode_collection.py]: This got changed with the QNode refactor :) After the refactor, `None` means no interface, which allows the QNode to be used without any autograd/Torch/TF overhead. Useful in the rare case where you don't care about autodifferentiation (gradients can still be called manually using `qnode1.jacobian()`).

File: pennylane/collections.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/collections.py]: @mariaschuld what was the error message you got?

File: pennylane/collections.py (hunk 2 above)
Comment (josh146) [reply pennylane/collections.py]: :+1: I'll move it to the top as per Maria's suggestion

**Comment by josh146** [comment]
> And how does dask determine the number of cores etc? Can the user configure this? Should we add it to the docstring?

@mariaschuld I looked into adding this, but it's more complex than I thought, so I'm in favour of making a ticket and leaving this for the future. If everyone agrees, I think this is ready to be merged? @trbromley, just waiting for your approval on the documentation changes :)

**Comment by co9olguy** [comment]
:fireworks:
